ch32fun/ch32fun.h

```diff
@@ -817,11 +825,41 @@ extern "C" {
 // and take two cycles, so you typically would use 0, 2, 4, etc.
 #define ADD_N_NOPS( n ) asm volatile( ".rept " #n "\nc.nop\n.endr" );
 
+#define FUN_HIGH 0x1
+#define FUN_LOW 0x0
+#if defined(CH59x)
+#define GPIOA_ResetBits(pin)      (R32_PA_CLR |= pin)
+#define GPIOA_SetBits(pin)        (R32_PA_OUT |= pin)
+#define GPIOA_InverseBits(pin)    (R32_PA_OUT ^= pin)
+#define GPIOA_ReadPortPin(pin)    (R32_PA_PIN & (pin))
+#define GPIOB_ResetBits(pin)      (R32_PB_CLR |= pin)
+#define GPIOB_SetBits(pin)        (R32_PB_OUT |= pin)
+#define GPIOB_InverseBits(pin)    (R32_PB_OUT ^= pin)
+#define GPIOB_ReadPortPin(pin)    (R32_PB_PIN & (pin))
+#define GPIO_ResetBits(pin)       { if(pin & PB) GPIOB_ResetBits(pin); else GPIOA_ResetBits(pin); }
+#define GPIO_SetBits(pin)         { if(pin & PB) GPIOB_SetBits(pin); else GPIOA_SetBits(pin); }
+#define GPIO_InverseBits(pin)     { if(pin & PB) GPIOB_InverseBits(pin); else GPIOA_InverseBits(pin); }
+#define funDigitalRead(pin)       ( (pin & PB) ? GPIOB_ReadPortPin(pin) : GPIOA_ReadPortPin(pin) )
+#define funDigitalWrite( pin, value ) { if(value==FUN_HIGH){GPIO_SetBits(pin);} else if(value==FUN_LOW){GPIO_ResetBits(pin);} }
+
+#define GPIO_ModeCfg(pd_drv, pu, dir, pin, mode) { switch(mode) { \
```

Comment:
This is complicated enough, it should be a function (even if a static function).

Reply:
(Otherwise the code copy pasta will be big)

UNLESS! Did you test it?  Does the compiler compile it down?

Can it do that for switches?  Or only if's?

Reply:
I tested this, but do you mean check the compiler output assembly? That I did not test. I moved this to a macro following your comment, I thought you did not want this as a function. (or maybe I just got compiler errors having this as a function in the header, I actually do not recall)

Reply:
Looking at it again I actually find this a very elegant and readable solution. What is exactly your objection? Would it be acceptable if the compiler optimizes it out fully like this?

Reply:
ouch this is why I had them as macros, the linker complains about multiple definitions and if I make them static or inline (or both) it actually does not work at all! Any suggestions how to resolve this?